eval/topdown.go

```diff
@@ -540,6 +588,174 @@ func evalRefRec(ctx *TopDownContext, ref opalog.Ref, iter TopDownIterator, path
 	}
 }
 
+func evalRefRule(ctx *TopDownContext, ref opalog.Ref, path opalog.Ref, rule *opalog.Rule, iter TopDownIterator) error {
+
+	switch rule.DocKind() {
+	case opalog.PartialSetDoc:
+		return evalRefRulePartialSetDoc(ctx, ref, path, rule, iter)
+	case opalog.PartialObjectDoc:
+		return evalRefRulePartialObjectDoc(ctx, ref, path, rule, iter)
+	case opalog.CompleteDoc:
+		return evalRefRuleCompleteDoc(ctx, ref, path, rule, iter)
+	default:
+		panic(fmt.Sprintf("illegal argument: %v", rule))
+	}
+
+}
+
+func evalRefRuleCompleteDoc(ctx *TopDownContext, ref opalog.Ref, path opalog.Ref, rule *opalog.Rule, iter TopDownIterator) error {
+	suffix := ref[len(path):]
+	if len(suffix) == 0 {
+		return fmt.Errorf("not implemented: %v %v %v", ref, path, rule)
+	}
+
+	bindings := newHashMap()
+	child := ctx.Child(rule, bindings)
+
+	return TopDown(child, func(child *TopDownContext) error {
+		switch v := rule.Value.Value.(type) {
+		case opalog.Object:
+			return evalRefRuleResult(ctx, ref, suffix, v, iter)
+		case opalog.Array:
+			return evalRefRuleResult(ctx, ref, suffix, v, iter)
+		default:
+			return fmt.Errorf("cannot dereference value (%T) in %v", rule.Value.Value, rule)
+		}
+	})
+}
+
+func evalRefRulePartialObjectDoc(ctx *TopDownContext, ref opalog.Ref, path opalog.Ref, rule *opalog.Rule, iter TopDownIterator) error {
+	suffix := ref[len(path):]
+	if len(suffix) == 0 {
+		return fmt.Errorf("not implemented: %v %v %v", ref, path, rule)
+	}
+
+	if !suffix[0].IsGround() {
+		child := ctx.Child(rule, newHashMap())
+		return TopDown(child, func(child *TopDownContext) error {
+			key := child.Bindings.Get(rule.Key.Value)
+			if key == nil {
+				return fmt.Errorf("unbound variable: %v", rule.Key)
+			}
+			value := child.Bindings.Get(rule.Value.Value)
+			if value == nil {
+				return fmt.Errorf("unbound variable: %v", rule.Value)
+			}
+			ctx = ctx.BindVar(suffix[0].Value.(opalog.Var), key)
+			return evalRefRuleResult(ctx, ref, ref[len(path)+1:], value, iter)
+		})
+	}
+
+	bindings := newHashMap()
+	bindings.Put(rule.Key.Value, suffix[0].Value)
+	child := ctx.Child(rule, bindings)
+
```

Comment:
Just had this discussion in person, but for posterity... We might want to restructure this later so that (i) plug context into ref, (ii) recurse, (iii) pull bindings from variables remaining in plugged ref and add to current binding